django/thunderstore/social/providers.py

```diff
@@ -0,0 +1,112 @@
+from typing import Dict
+
+import requests
+from django.conf import settings
+
+
+class BaseOauthHelper:
+    token = None
+
+    def __init__(self, code: str, redirect_url: str) -> None:
+        self.code = code
+        self.redirect_url = redirect_url
+
+    def complete_login(self) -> None:
+        """
+        Fetch access token with the temporary code and configured secret.
+        """
+        data = {
+            "client_id": self.CLIENT_ID,
+            "client_secret": self.CLIENT_SECRET,
+            "code": self.code,
+            "grant_type": "authorization_code",  # Required by Discord
+            "redirect_uri": self.redirect_url,  # Required by Discord
+        }
+        headers = {"Accept": "application/json"}
+
+        response = requests.post(self.OAUTH_URL, data, headers=headers)
+        response_data = response.json()
+        self.token = response_data["access_token"]
+
+    def get_user_info(self) -> Dict:
+        raise NotImplementedError()
+
+    def _fetch_from_api(self, url: str) -> Dict:
+        """
+        Use access token to fetch info from provider's public API.
+        """
+        if self.token is None:
+            raise Exception("No token found. Did you call .complete_login()?")
```

Comment:
Perhaps we could make the token a required parameter in the function instead of assuming the state of the object is correct? Works better with static analysis that way around and in theory makes the whole check unnecessary (in practice the check is good to have because we don't have a type linter in the CI pipeline yet).

Reply:
If this method treats `token` as a required (`str` instead of `Optional[str]`) parameter, then we'd need to move (and duplicate) the check in all the places where this method is called? Otherwise the static analysis complains when we're trying to pass `Optional[str]` to method that accepts only `str`? Or am I missing something here?

Reply:
That's correct. So perhaps that'd generate a lot of boilerplate, we could add a standalone method for validating a token, which in terms of type-declarations converts `Optional[str]` to `str` or raises the appropriate error? Mainly I'd see this approach "cleaner" from a static analysis & self-documenting code perspective, as it should be immediately obvious the method needs a token to exist.

Reply:
Implemented as suggested.